### ISSUE TITLE
Refactor package structure

### ISSUE THREAD

**Issue by luca-moser** [iotaledger/goshimmer#341]
In Go, it is idiomatic to have as many file as one wants within one package, doing many sub packages is usually not idiomatic. My gut feeling momentarily is that we have too many sub packages with too many duplicates in naming. For example, have a look at  the [`net/http` package from the stdlib](https://github.com/golang/go/tree/master/src/net/http), it exposes cookie, request/response, client and server functionality all from within the same package.

I believe we can structure layers related logic in the following way:
* `/pkg/layer/message` 
* `/pkg/layer/drng`
* `/pkg/layer/value`

Taking `/pkg/layer/message` as an example, it would contain files like `factory.go`, `requester.go`, `filter.go`, `tip_selector.go` etc. all in one directory. Instead of having them scattered in sub packages.

Then the obvious question arises with things like `events.go` which seem to exist in every sub package. We can either rename them to be something like `FactoryEvents`, `RequesterEvents` etc. or simply inline the struct into the corresponding `Factory` / `Requester` structs. This way we gain the ability to either query for `FactoryEvents` or we can simply see the events by inspecting the actual struct. In fact, most `Events` structs are solely defined but ultimately end up being a field of another struct anyway right now.

With a flatter package structure, it is clearer what you're dealing with. For example, instead of having `/packages/binary/messagelayer/messagefactory.MessageFactory` (which includes the word "message" way too many times), one can simply import `/pkg/layer/message` and you would have `message.Factory` and it is automatically clear that you have a factory to make base messages. If you were to deal with value message, you'd end up with something like `valueMsg := &value.Message{}` which is also clear.

**Comment by hmoog** [comment]
I disagree.

The goshimmer repository is not a classical go package which exposes a single self-contained "library" with a common logic but a "monorepo", containing packages and plugins for multiple different self-contained building blocks of our node framework (together with the node framework itself). Instead of creating a separate repository for every single building block on github, we store and develop all of the different building blocks as part of the same git repository.

Every decentralized application/module should therefore define its own go idiomatic package structure, which then follows go idiomatic naming of the sub packages. The "containing" folder that wraps all of the relevant sub-packages into a single folder should simply be a "container" in the github repository representing the "sub-project". But it is not supposed to be understood as part of the same hierarchy of packages.

IMHO, your proposed folder and package structure is also misleading:

`/pkg/layer/message/factory.go` would be a MessageLayerFactory and not a MessageFactory

We should instead think about a package and folder structure, that correctly reflects the monorepo aspect, by putting all the logic that belongs to a certain sub-project into its own container i.e.

- `/dapps/<dappName>/packages/...`
- `/dapps/<dappName>/plugins/...`
- `/dapps/<dappName>/<dappName.go>`

This will also allow people to install new application by simply copying a single folder into their local  filesystem and enable the corresponding plugin by adding it to their main.go.

TL;DR: If the goshimmer repo would be a classical go package exposing a single self contained library, then I would agree - but since its a mono-repo and we are losing the ability to simply copy in a single folder using your proposed package layout, i think it does not fit.

**Comment by luca-moser** [comment]
Adjust packages for:
* message layer
* value transfer layer
* faucet
